File: Sources/custom_adapter/TopOnAdapter/Objc/6.5.x/Banner/AlxTopOnBannerAdapter.m
//
//  AlxTopOnBannerAdapter.m
//  AlxAdsOCDemo
//

#import "AlxTopOnBannerAdapter.h"
#import "AlxTopOnBaseManager.h"
#import "AlxTopOnBannerEvent.h"
#import "AlxTopOnBiddingRequest.h"
#import "AlxTopOnBiddingRequestManager.h"
#import "AlxTopOnTool.h"
#import "AlxTopOnBannerDelegate.h"

static NSString *const TAG = @"AlxTopOnBannerAdapter";

/**
 * 告诉编译器：这个类其实是会这个协议的。
 * Tell the compiler that this class actually conforms to this protocol.
 */
@interface ATBannerAdStatusBridge (AlxDelegate) <AlxBannerViewAdDelegate>
@end

@interface AlxTopOnBannerAdapter ()
/**
 * Alx SDK 的 banner 广告对象。
 * Alx SDK banner ad object.
 */
@property (nonatomic, strong) AlxBannerAdView *bannerAd;
// delegate
@property (nonatomic, strong) AlxTopOnBannerDelegate *bannerDelegate;
@end

@implementation AlxTopOnBannerAdapter

#pragma mark - lazy load
/**
 * 初始化bannerDelegate属性。
 * Initialize the bannerDelegate property.
 */
- (AlxTopOnBannerDelegate *)bannerDelegate {
    if (_bannerDelegate == nil) {
        _bannerDelegate = [[AlxTopOnBannerDelegate alloc] init];
        _bannerDelegate.adStatusBridge = self.adStatusBridge;
    }
    return _bannerDelegate;
}

#pragma mark - Ad load
/**
 * 实现广告加载方法。
 * Implement the ad loading method.
 */
- (void)loadADWithArgument:(ATAdMediationArgument *)argument {
    NSLog(@"%@: loadAD", TAG);
    NSLog(@"%@: loadAD: isMainThread=%@", TAG, [NSThread isMainThread] ? @"YES" : @"NO");
    
    NSString *bidId = argument.serverContentDic[kATAdapterCustomInfoBuyeruIdKey];
    NSLog(@"%@: loadAD: bidId=%@", TAG, bidId ?: @"空");
    
    // 文档要求，通常需要在主线程执行 / As required by the documentation, this usually needs to run on the main thread
    dispatch_async(dispatch_get_main_queue(), ^{
        NSString *uuitKeyStr = [AlxTopOnBaseManager unitID];
        NSString *uuitId = argument.serverContentDic[uuitKeyStr];
        if (!uuitId || uuitId.length == 0) {
            NSString *errorStr = @"uuitid is empty";
            NSLog(@"%@: loadAD: error = %@", TAG, errorStr);
            [AlxTopOnBaseManager errorWithCode:-100 message:errorStr];
            return;
        }
        NSLog(@"%@: loadAD: uuitid = %@", TAG, uuitId);
        
        if (bidId) {
            // Bidding 场景：从缓存中取出已加载的广告 / Bidding scenario: retrieve the pre-loaded ad from cache
            AlxTopOnBiddingRequest *biddingRequest = [[AlxTopOnTool shared] getRequestItemWithUnitID:uuitId];
            if (biddingRequest) {
                self.bannerAd = (AlxBannerAdView *)biddingRequest.customObject;
                
                if (self.bannerAd) {
                    NSLog(@"%@: loadAD: bid ad loaded, notify success", TAG);
                    [self.adStatusBridge atOnBannerAdLoadedWithView:self.bannerAd adExtra:nil];
                } else {
                    NSLog(@"%@: loadAD: bid ad object is empty", TAG);
                    NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid ad object is empty"}];
                    [self.adStatusBridge atOnAdLoadFailed:error adExtra:nil];
                }
            } else {
                NSLog(@"%@: loadAD: bid request not found in cache", TAG);
                NSError *error = [NSError errorWithDomain:@"AlxTopOnAdapter" code:-100 userInfo:@{NSLocalizedDescriptionKey: @"Bid request not found"}];
                [self.adStatusBridge atOnAdLoadFailed:error adExtra:nil];
            }
            // ⚠️ 注意：移除缓存 / ⚠️ Note: Remove from cache
            [[AlxTopOnTool shared] removeRequestItemWithUnitID:uuitId];
        } else {
            ATUnitGroupModel *unitGroupModel = argument.serverContentDic[kATAdapterCustomInfoUnitGroupModelKey];
            // 通过argument对象获取必要的加载信息，如尺寸等，创建好必要的参数，准备传入给第三方的横幅加载方法，开始加载广告
            // Get necessary loading info (e.g. size) from the argument object, prepare parameters, and pass them to the third-party banner loading method to start loading the ad
            CGSize bannerSize = CGSizeMake(320, 50);
            if (!CGSizeEqualToSize(argument.bannerSize, CGSizeZero) && unitGroupModel) {
                bannerSize = argument.bannerSize;
                NSLog(@"%@: loadAD: width = %.2f, height = %.2f", TAG, bannerSize.width, bannerSize.height);
            }
            
            self.bannerAd = [[AlxBannerAdView alloc] initWithFrame:CGRectMake(0, 0, bannerSize.width, bannerSize.height)];
            
            // ⚠️ 注意：delegate 应该设置为 bannerDelegate，而不是 adStatusBridge
            // ⚠️ Note: delegate should be set to bannerDelegate, not adStatusBridge
            self.bannerAd.delegate = self.bannerDelegate;
            self.bannerAd.refreshInterval = 0;
            self.bannerAd.translatesAutoresizingMaskIntoConstraints = NO;
            
            NSLog(@"%@: start loading ad with unitId: %@", TAG, uuitId);
            [self.bannerAd loadAdWithAdUnitId:uuitId];
        }
    });
}


//- (instancetype)initWithNetworkCustomInfo:(NSDictionary *)serverInfo localInfo:(NSDictionary *)localInfo {
//    self = [super init];
//    if (self) {
//        NSLog(@"%@: init", TAG);
//        NSLog(@"%@: init: isMainThread=%@", TAG, [NSThread isMainThread] ? @"YES" : @"NO");
//        dispatch_async(dispatch_get_main_queue(), ^{
//            if (![AlxTopOnBaseManager isInitialized]) {
//                [AlxTopOnBaseManager initSDKWithServerInfo:serverInfo];
//            }
//        });
//    }
//    return self;
//}

+ (void)bidRequestWithPlacementModel:(ATPlacementModel *)placementModel
                      unitGroupModel:(ATUnitGroupModel *)unitGroupModel
                                info:(NSDictionary *)info
                          completion:(void (^)(ATBidInfo * _Nullable, NSError * _Nullable))completion {
    NSLog(@"%@: bidRequestWithPlacementModel", TAG);
    
    dispatch_async(dispatch_get_main_queue(), ^{
        if (![AlxTopOnBaseManager isInitialized]) {
            [AlxTopOnBaseManager initSDKWithServerInfo:info];
        }
    });
    
    AlxTopOnBannerEvent *customEvent = [[AlxTopOnBannerEvent alloc] initWithInfo:info localInfo:info];
    customEvent.isC2SBiding = YES;
    
    AlxTopOnBiddingRequest *request = [[AlxTopOnBiddingRequest alloc]
                                       initWithUnitGroup:unitGroupModel
                                       customEvent:customEvent
                                       unitID:info[[AlxTopOnBaseManager unitID]]
                                       placementID:placementModel.placementID
                                       extraInfo:info
                                       adType:ATAdFormatBanner
                                       bidCompletion:completion];
    
    [[AlxTopOnBiddingRequestManager shared] startWithRequest:request];
}

+ (BOOL)adReadyWithCustomObject:(id)customObject info:(NSDictionary *)info {
    NSLog(@"%@: adReady", TAG);
    if ([customObject isKindOfClass:[AlxBannerAdView class]]) {
        NSLog(@"%@: adReady true", TAG);
        return YES;
    } else {
        NSLog(@"%@: adReady false", TAG);
        return NO;
    }
}

+ (void)showBanner:(ATBanner *)banner inView:(UIView *)view presentingViewController:(UIViewController *)viewController {
    NSLog(@"%@: showBanner", TAG);
    AlxBannerAdView *bannerView = banner.customObject;
    if (!bannerView) {
        NSLog(@"%@: showBanner banner is nil", TAG);
        return;
    }
    
    bannerView.rootViewController = viewController;
    for (UIView *subview in view.subviews) {
        [subview removeFromSuperview];
    }
    [view addSubview:bannerView];
}

@end

File: custom_adapter/TopOnAdapter/AlxTopOnBannerAdapter.swift
//
//  AlxTopOnBannerAdapter.swift
//

import Foundation
import AnyThinkSDK
import AnyThinkBanner
import AlxAds

@objc(AlxTopOnBannerAdapter)
public class AlxTopOnBannerAdapter: ATAdAdapter {
    
    private static let TAG = "AlxTopOnBannerAdapter"
    
    private var bannerAd:AlxBannerAdView? = nil
    private var customEvent:AlxTopOnBannerEvent? = nil
    
    // MARK: - Initialization
    @objc public required init(networkCustomInfo serverInfo: [AnyHashable: Any],
                        localInfo: [AnyHashable: Any]) {
        super.init()
        NSLog("%@: init",AlxTopOnBannerAdapter.TAG)
        NSLog("%@: init: isMainThread=%@",AlxTopOnBannerAdapter.TAG,Thread.current.isMainThread ? "YES" : "NO")
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: serverInfo)
            }
        }
    }
    
    /// 加载广告
    @objc public func loadAD(withInfo serverInfo: [AnyHashable: Any],
                             localInfo: [AnyHashable: Any],
                             completion: @escaping ([[AnyHashable: Any]]?, (any Error)?) -> Void) {
        NSLog("%@: loadAD",AlxTopOnBannerAdapter.TAG)
        NSLog("%@: loadAD: isMainThread=%@",AlxTopOnBannerAdapter.TAG,Thread.current.isMainThread ? "YES" : "NO")
        
        let bidId:String? = serverInfo[kATAdapterCustomInfoBuyeruIdKey] as? String
        NSLog("%@: loadAD: bidId=%@",AlxTopOnBannerAdapter.TAG,bidId ?? "空")
        
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: serverInfo)
            }
            
            let bidId=serverInfo[kATAdapterCustomInfoBuyeruIdKey] as? String
            NSLog("%@: loadAD: bidId=%@",AlxTopOnBannerAdapter.TAG,bidId ?? "")
            
            guard let unitId = serverInfo[AlxTopOnBaseManager.unitID] as? String,!unitId.isEmpty else{
                let errorStr="unitid is empty"
                NSLog("%@: loadAD: error= %@",AlxTopOnBannerAdapter.TAG,errorStr)
                completion(nil,AlxTopOnBaseManager.error(code: -100, msg: errorStr))
                return
            }
            
            NSLog("%@: loadAD: unitid=%@",AlxTopOnBannerAdapter.TAG,unitId)
            
            if bidId != nil {
                let request:AlxTopOnBiddingRequest? = AlxTopOnTool.shared.getRequestItem(withUnitID: unitId) as? AlxTopOnBiddingRequest
                if let request = request {
                    self.customEvent = request.customEvent as? AlxTopOnBannerEvent
                    self.customEvent?.requestCompletionBlock = completion
                    self.bannerAd = request.customObject as? AlxBannerAdView
                    //判断广告源是否已经loaded过
                    if let bannerAd = self.bannerAd {
                        self.customEvent?.trackBannerAdLoaded(bannerAd, adExtra: nil)
                    }else{
                        NSLog("%@: loadAD: bid ad object is empty",AlxTopOnBannerAdapter.TAG)
                    }
                }
                AlxTopOnTool.shared.removeRequestItem(withUnitID: unitId)
            }else{
                
                self.customEvent = AlxTopOnBannerEvent(info: serverInfo, localInfo: localInfo)
                self.customEvent?.requestCompletionBlock = completion
                
                // 提取广告尺寸信息
                var adSize = CGSize(width: 320, height: 50)
                let unitGroupModel:ATUnitGroupModel? = serverInfo[kATAdapterCustomInfoUnitGroupModelKey] as? ATUnitGroupModel
                if let unitGroupModel = unitGroupModel {
                    adSize = unitGroupModel.adSize
                    NSLog("%@: loadAD: width=%.2f , height=%.2f",AlxTopOnBannerAdapter.TAG,adSize.width,adSize.height)
                }
                
                // load ad
                self.bannerAd=AlxBannerAdView(frame: CGRect(origin: .zero, size: adSize))
                self.bannerAd?.delegate = self.customEvent
                self.bannerAd?.refreshInterval = 0
                self.bannerAd?.translatesAutoresizingMaskIntoConstraints = false
                self.bannerAd?.loadAd(adUnitId: unitId)
            }
        }
        
    }
    
    // MARK: - C2S header bidding 竞价
    @objc public static func bidRequestWithPlacementModel(_ placementModel: ATPlacementModel,unitGroupModel: ATUnitGroupModel, info: [AnyHashable: Any], completion: @escaping (ATBidInfo?, Error?) -> Void) {
        NSLog("%@: bidRequestWithPlacementModel",AlxTopOnBannerAdapter.TAG)
        
        DispatchQueue.main.async {
            if !AlxTopOnBaseManager.isInitialized {
                AlxTopOnBaseManager.initSDK(serverInfo: info)
            }
        }
        
        let customEvent:AlxTopOnBannerEvent = AlxTopOnBannerEvent(info: info, localInfo: info)
        customEvent.isC2SBiding = true
        let request:AlxTopOnBiddingRequest = AlxTopOnBiddingRequest(
            unitGroup: unitGroupModel,
            customEvent: customEvent,
            unitID: info[AlxTopOnBaseManager.unitID] as? String,
            placementID: placementModel.placementID,
            extraInfo: info,
            adType: ATAdFormat.banner,
            bidCompletion: completion
        )
        AlxTopOnBiddingRequestManager.shared.start(with: request)
    }
    
    
    // MARK: - 广告就绪检查
    @objc public static func adReady(withCustomObject customObject: Any, info: [AnyHashable: Any]) -> Bool {
        // 检查广告是否就绪
        // 实际实现中应根据广告网络的具体方法进行检查
        NSLog("%@: adReady",AlxTopOnBannerAdapter.TAG)
        if customObject as? AlxBannerAdView != nil {
            NSLog("%@: adReady true",AlxTopOnBannerAdapter.TAG)
            return true
        }else{
            NSLog("%@: adReady false",AlxTopOnBannerAdapter.TAG)
            return false
        }
    }
    
    // MARK: - 广告展示
    @objc(showBanner:inView:presentingViewController:)
    public static func showBanner(_ banner: ATBanner, in view: UIView, presenting viewController: UIViewController) {
        // 清空容器视图
        NSLog("%@: showBanner",AlxTopOnBannerAdapter.TAG)
        guard let bannerView = (banner.customObject as? AlxBannerAdView) else {
            NSLog("%@: showBanner banner is nil",AlxTopOnBannerAdapter.TAG)
            return
        }
        bannerView.rootViewController = viewController
        view.subviews.forEach { $0.removeFromSuperview() }
        view.addSubview(bannerView)
    }
    
}
